File: java/Tree.java
import java.util.ArrayList;
import java.util.List;

class Tree {
  int n;
  int[] u;
  int[] v;
  List<Integer>[] edgeLists;
  int[] depths;
  int[][] ancestors;

  @SuppressWarnings("unchecked")
  Tree(int[] u, int[] v) {
    n = u.length + 1;

    this.u = u;
    this.v = v;

    edgeLists = new List[n];
    for (int i = 0; i < edgeLists.length; ++i) {
      edgeLists[i] = new ArrayList<>();
    }
    for (int i = 0; i < u.length; ++i) {
      edgeLists[u[i]].add(i);
      edgeLists[v[i]].add(i);
    }

    depths = new int[n];
    ancestors = new int[n][Integer.toBinaryString(n).length()];
    init(0, -1, 0);
  }

  private void init(int depth, int parent, int node) {
    depths[node] = depth;

    ancestors[node][0] = parent;
    for (int i = 1; i < ancestors[node].length; ++i) {
      ancestors[node][i] =
          (ancestors[node][i - 1] == -1) ? -1 : ancestors[ancestors[node][i - 1]][i - 1];
    }

    for (int edge : edgeLists[node]) {
      int adj = (node == u[edge]) ? v[edge] : u[edge];
      if (adj != parent) {
        init(depth + 1, node, adj);
      }
    }
  }

  int findLca(int node1, int node2) {
    if (depths[node1] < depths[node2]) {
      return findLca(node2, node1);
    }

    for (int i = ancestors[node1].length - 1; i >= 0; --i) {
      if (ancestors[node1][i] != -1 && depths[ancestors[node1][i]] >= depths[node2]) {
        node1 = ancestors[node1][i];
      }
    }

    if (node1 == node2) {
      return node1;
    }

    for (int i = ancestors[node1].length - 1; i >= 0; --i) {
      if (ancestors[node1][i] != ancestors[node2][i]) {
        node1 = ancestors[node1][i];
        node2 = ancestors[node2][i];
      }
    }

    return ancestors[node1][0];
  }
}

File: rust/src/tree.rs
#[allow(dead_code)]
struct Tree {
    n: usize,
    u: Vec<usize>,
    v: Vec<usize>,
    root: usize,
    edge_vecs: Vec<Vec<usize>>,
    depths: Vec<i32>,
    ancestors: Vec<Vec<usize>>,
}

#[allow(dead_code)]
impl Tree {
    fn new(u: &[usize], v: &[usize], root: usize) -> Self {
        let n = u.len() + 1;

        let mut edge_vecs = vec![Vec::new(); n];
        for i in 0..u.len() {
            edge_vecs[u[i]].push(i);
            edge_vecs[v[i]].push(i);
        }

        let mut depths = vec![0; n];
        let mut ancestors = vec![vec![usize::MAX; (n.ilog2() as usize) + 1]; n];
        Self::init(
            &mut depths,
            &mut ancestors,
            u,
            v,
            &edge_vecs,
            0,
            usize::MAX,
            root,
        );

        Self {
            n,
            u: u.to_vec(),
            v: v.to_vec(),
            root,
            edge_vecs,
            depths,
            ancestors,
        }
    }

    fn init(
        depths: &mut [i32],
        ancestors: &mut [Vec<usize>],
        u: &[usize],
        v: &[usize],
        edge_vecs: &[Vec<usize>],
        depth: i32,
        parent: usize,
        node: usize,
    ) {
        depths[node] = depth;

        ancestors[node][0] = parent;
        for i in 1..ancestors[node].len() {
            if ancestors[node][i - 1] != usize::MAX {
                ancestors[node][i] = ancestors[ancestors[node][i - 1]][i - 1];
            }
        }

        for &edge in &edge_vecs[node] {
            let adj = if node == u[edge] { v[edge] } else { u[edge] };

            if adj != parent {
                Self::init(depths, ancestors, u, v, edge_vecs, depth + 1, node, adj);
            }
        }
    }

    fn find_lca(&self, mut node1: usize, mut node2: usize) -> usize {
        if self.depths[node1] < self.depths[node2] {
            return self.find_lca(node2, node1);
        }

        for i in (0..self.ancestors[node1].len()).rev() {
            if self.ancestors[node1][i] != usize::MAX
                && self.depths[self.ancestors[node1][i]] >= self.depths[node2]
            {
                node1 = self.ancestors[node1][i];
            }
        }

        if node1 == node2 {
            return node1;
        }

        for i in (0..self.ancestors[0].len()).rev() {
            if self.ancestors[node1][i] != self.ancestors[node2][i] {
                node1 = self.ancestors[node1][i];
                node2 = self.ancestors[node2][i];
            }
        }

        self.ancestors[node1][0]
    }
}
